### PR TITLE
Use cluster view to replace list of workers

### PR DESCRIPTION
**Pull request by dbw9580**
### What changes are proposed in this pull request?

Replace `List<BlockWorkerInfo>` with `WorkerClusterView` in APIs.

Important APIs that are changed:

1. `FileSystemContext.getCachedWorkers` now returns `WorkerClusterView`
2. `WorkerLocationPolicy.getPreferredWorkers` (as well as all its implementations) now accepts a `WorkerClusterView` as the first argument (but still returns `List<BlockWorkerInfo>` as the returned list must be ordered)

APIs that are using `List<BlockWorkerInfo>` (or `List<WorkerInfo>`) but *not* migrated to `WorkerClusterView`:

1. `alluxio.master.scheduler.WorkerProvider.getWorkerInfos` returns `List<WorkerInfo>`.
2. Job service related APIs, e.g. `alluxio.job.plan.PlanDefinition.selectExecutors`

Notable behavior change:

Now `EtcdMembershipManager` assigns the correct state (`LIVE` or `LOST`) for all workers in its  `WorkerInfo` struct. Before this change, this information is not available and the state defaults to `UNRECOGNIZED`.

### Why are the changes needed?

Allow more efficient indexing and filtering workers by worker ID.

### Does this PR introduce any user facing changes?

No.

### CODE REVIEW

**Comment by dbw9580** [comment]
@lucyge2022 @JiamingMai @jiacheliu3 can you please take a look? thanks.

@jja725 I left the scheduler related APIs unchanged but I think it can also benefit from this refactor. Let me know what you think.

File: dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java [hunk 1]
```diff
@@ -116,11 +118,17 @@ public void join(WorkerInfo workerInfo) throws IOException {
 
   @Override
   public WorkerClusterView getAllMembers() throws IOException {
+    Set<WorkerIdentity> liveWorkerIds = parseWorkersFromEtcdKvPairs(
+        mAlluxioEtcdClient.mServiceDiscovery.getAllLiveServices())
+        .map(WorkerServiceEntity::getIdentity)
+        .collect(Collectors.toSet());
+    Predicate<WorkerInfo> isLive = w -> liveWorkerIds.contains(w.getIdentity());
```
Comment (dbw9580) [inline dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java:125]: to figure out the liveness of a worker, I had to first get a set of all live workers, so that live and lost workers can be correctly differentiate. @lucyge2022 I'd appreciate your comment on this.

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 2]
```diff
@@ -947,50 +947,26 @@ public List<BlockWorkerInfo> getLostWorkers() throws IOException {
...
-        List<BlockWorkerInfo> liveWorkers = mMembershipManager.getLiveMembers().stream()
-            .map(w -> new BlockWorkerInfo(w.getIdentity(), w.getAddress(), w.getCapacityBytes(),
-                w.getUsedBytes(), true)).collect(toList());
-        List<BlockWorkerInfo> lostWorkers = mMembershipManager.getFailedMembers().stream()
-            .map(w -> new BlockWorkerInfo(
-                w.getIdentity(), w.getAddress(), w.getCapacityBytes(), w.getUsedBytes(),
-                false)).collect(toList());
-        // avoid duplicate elements in list
-        return combineAllWorkers(liveWorkers, lostWorkers);
+        return mMembershipManager.getAllMembers();
```
Comment (dbw9580) [inline dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:956]: I was confused why we are taking trouble to combine a list of live workers with a list of lost workers to get the list of all workers, while all this time there is a `getAllMembers` method to do exactly this.

I think this change has a subtle difference, as the `combineAllWorkers` method used to de-duplicate workers by their net address. Now `mMembershipManager.getAllMembers` does so by checking the worker IDs. But workers shouldn't have conflicting net addresses either so it's no big problem IMO.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java [hunk 3]
```diff
@@ -50,14 +53,18 @@ public ConsistentHashPolicy(AlluxioConfiguration conf) {
...
           "Not enough workers in the cluster %d workers in the cluster but %d required",
-          blockWorkerInfos.size(), count));
+          workerClusterView.size(), count));
     }
-    HASH_PROVIDER.refresh(blockWorkerInfos, mNumVirtualNodes);
+    List<BlockWorkerInfo> blockWorkerInfoList = workerClusterView.stream()
+        .map(w -> new BlockWorkerInfo(w.getIdentity(), w.getAddress(), w.getCapacityBytes(),
+            w.getUsedBytes(), w.getState() == WorkerState.LIVE))
+        .collect(Collectors.toList());
+    HASH_PROVIDER.refresh(blockWorkerInfoList, mNumVirtualNodes);
```
Comment (dbw9580) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:67]: refactoring `ConsistentHashProvider` to accept `WorkerClusterView` instead of `List<BlockWorkerInfo>` will be done in a separate PR: https://github.com/Alluxio/alluxio/pull/18434

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 4]
```diff
@@ -39,17 +40,17 @@ public interface WorkerLocationPolicy {
...
    *
-   * @param blockWorkerInfos
+   * @param workers
    * @param fileId
    * @param count
    * @return a list of preferred workers
    * @throws ResourceExhaustedException if unable to return exactly #{count} workers
    */
-  List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWorkerInfos,
+  List<BlockWorkerInfo> getPreferredWorkers(WorkerClusterView workers,
```
Comment (dbw9580) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:53]: the return type is unchanged because the order of the workers returned is important.

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 5]
```diff
@@ -178,15 +175,22 @@ public class FileSystemContext implements Closeable {
...
+   */
   private boolean mUriValidationEnabled = true;
 
-  /** Cached map for workers. */
-  @GuardedBy("mWorkerInfoList")
-  private final AtomicReference<List<BlockWorkerInfo>> mWorkerInfoList = new AtomicReference<>();
+  /**
+   * Cached map for workers.
+   */
+  @GuardedBy("this")
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:186]:
```suggestion
  @GuardedBy("mCachedWorkerClusterView")
```
The `this` in the class is actually the `FileSystemContext` object.

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 6]
```diff
@@ -869,24 +874,24 @@ public List<BlockWorkerInfo> getCachedWorkers() throws IOException {
...
-    synchronized (mWorkerInfoList) {
-      if (mWorkerInfoList.get() == null || mWorkerInfoList.get().isEmpty()
+  public WorkerClusterView getCachedWorkers(GetWorkerListType type) throws IOException {
+    synchronized (mCachedWorkerClusterView) {
+      if (mCachedWorkerClusterView.get() == null || mCachedWorkerClusterView.get().isEmpty()
           || mWorkerRefreshPolicy.attempt()) {
         switch (type) {
           case ALL:
-            mWorkerInfoList.set(getAllWorkers());
+            mCachedWorkerClusterView.set(getAllWorkers());
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:883]: oh interestingly, this same cache object is caching results for 3 types? how do we know if a get wants the same type as is cached?

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 2 above)
Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:956]: Agreed we can use `mMembershipManager.getAllMembers` given the subtle difference. But the property that `workers shouldn't have conflicting net addresses` is not guaranteed IMO? I suggest we double check that in the worker register code (in a separate PR if necessary). At least if some workers are using dup IDs or addresses we should check somewhere and log some warnings, so that people can realize at all.

**Review by jiacheliu3 — APPROVED** [review]
Mostly LGTM. Good to merge once the comments are looked at. Thanks for the work!

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 6 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:883]: good catch! I think this is a potential bug. Since the only caller to this overloaded `getCachedWorkers(GetWorkerListType)` is `DoraCacheClient.getWorkerNetAddress` and it's argument `GetWorkerListType` is tied to `mEnableDynamicHashRing` which is currently a runtime constant, this does not cause any visible bug.

IMO since there are already `getLiveWorkers` `getAllWorkers` and `getLostWorkers` methods on `FileSystemContext`, I'd propose to simply deprecate/remove this method.

Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:883]: to fix this in a separate PR.

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 2 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:956]: > workers shouldn't have conflicting net addresses is not guaranteed

Right. So there used to be some potential bugs due to conflicting worker addresses. But this is not relevant any more. Depending on net addresses to differentiate workers is not reliable, hence the introduction of worker IDs.

> we double check that in the worker register code  

I think Lucy did exactly that in https://github.com/Alluxio/alluxio/pull/18454

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 5 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: done

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 7]
```diff
@@ -178,15 +175,22 @@ public class FileSystemContext implements Closeable {
...
+  @GuardedBy("itself")
+  private final AtomicReference<WorkerClusterView> mCachedWorkerClusterView =
+      new AtomicReference<>();
 
-  /** The policy to refresh workers list. */
-  @GuardedBy("mWorkerInfoList")
+  /**
+   * The policy to refresh workers list.
+   */
+  @GuardedBy("mCachedWorkerClusterView")
```
Comment (lucyge2022) [inline dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:193]: why is this policy guarded by the workerlist? there's no update to this policy? can u remove this if its a legacy thing

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 6 above)
Comment (lucyge2022) [reply dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:883]: good catch

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 2 above)
Comment (lucyge2022) [reply dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:956]: @jiacheliu3   @dbw9580  actually if one worker register with workerid1 and then remove its workeridentity file and restart, it will register as workerid2,  if no one removes workerid1 key from etcd, getAllMembers will have both worker entity bearing same addresses, but that's not sth code base should be guarding against, its a deployment issue. getAllMembers could be thought of as all the distinct members forming the ring.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java [hunk 8]
```diff
@@ -50,14 +53,18 @@ public ConsistentHashPolicy(AlluxioConfiguration conf) {
...
+    if (workerClusterView.size() < count) {
       throw new ResourceExhaustedException(String.format(
           "Not enough workers in the cluster %d workers in the cluster but %d required",
-          blockWorkerInfos.size(), count));
+          workerClusterView.size(), count));
     }
-    HASH_PROVIDER.refresh(blockWorkerInfos, mNumVirtualNodes);
+    List<BlockWorkerInfo> blockWorkerInfoList = workerClusterView.stream()
+        .map(w -> new BlockWorkerInfo(w.getIdentity(), w.getAddress(), w.getCapacityBytes(),
+            w.getUsedBytes(), w.getState() == WorkerState.LIVE))
```
Comment (lucyge2022) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:65]: i thought by the time we get to this func, the original blockWorkerInfos is already set to be either a LIVE view or an ALL view

File: dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java (hunk 1 above)
Comment (lucyge2022) [reply dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java:164]: yeah u have to make separate calls to know the state of a worker

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java (hunk 8 above)
Comment (lucyge2022) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:65]: oh its setting the mActive of BlockworkerInfo, never mind the above comment

**Review by lucyge2022 — APPROVED** [review]
LGTM other than some minor comments

File: dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 7 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:193]: `RefreshPolicy` objects are not thread safe, and `TimeoutRefresh.attempt` actually mutates its internal state. Therefore access to it must be serialized across different threads.

This policy object is only used in `getCachedWorkers` so it's fine to guard it with `mCachedWorkerClusterView`'s lock.

**Comment by dbw9580** [comment]
alluxio-bot, merge this please

**Comment by dbw9580** [comment]
alluxio-bot, merge this please
